plugins/woocommerce-blocks/assets/js/blocks/product-collection/edit/inspector-controls/hand-picked-products-control.tsx

```diff
@@ -36,13 +42,27 @@ function useProducts() {
 	);
 
 	useEffect( () => {
-		getProducts( {
-			selected: [],
-			queryArgs: {
-				// Fetch all products.
-				per_page: 0,
-			},
-		} ).then( ( results ) => {
+		// We take two strategies here because of internal logic of
+		// `getProducts` and `getProductsRequests` that skips request for
+		// `selected` products for small stores. So fetching products per user's input
+		// breaks selected items:
+		// 1. For large stores (>100 products) we fetch products as input changes AND
+		// `selected` products.
+		// 2. For small stores (<=100 products) we fetch all products just once.
+
+		const query = {
+			selected: isLargeCatalog ? selected.map( Number ) : [],
+			queryArgs: isLargeCatalog
+				? {
+						search,
+						// Limit search to 40 results. If results are not satisfying
+						// user needs to type more characters to get closer to actual
+						// product name.
+						per_page: 40,
+				  }
+				: {},
```

Comment:
Nitpick: Maybe we can pass `per_page: 0,` explicitly in query while calling `getProducts` function. It will make it more clear that we are fetching all the products. :slightly_smiling_face:
Otherwise, We rely on [this logic](https://github.com/woocommerce/woocommerce/blob/eb33beb2ffb3f26b9661afb1b98901d5899dae55/plugins/woocommerce-blocks/assets/js/editor-components/utils/index.js#L24), which may stop working in future if inside [getProductsRequests](https://github.com/woocommerce/woocommerce/blob/eb33beb2ffb3f26b9661afb1b98901d5899dae55/plugins/woocommerce-blocks/assets/js/editor-components/utils/index.js#L17-L21) function, [this condition](https://github.com/woocommerce/woocommerce/blob/eb33beb2ffb3f26b9661afb1b98901d5899dae55/plugins/woocommerce-blocks/assets/js/editor-components/utils/index.js#L22) will be changed i.e. instead of using 100 to calculate `isLargeCatalog`, it is changed to something else. 

Feel free to ignore this if you think it's not needed. Also, Please let me know if I misunderstood something.

Reply:
Yeah, I agree it makes sense to pass it explicitly! :) changed in https://github.com/woocommerce/woocommerce/pull/48379/commits/ae536960a1c600707543751d3f6c18c64c4de1cc